### PR TITLE
[SPARK-14437][Core]Use the address that NettyBlockTransferService listens to create BlockManagerId

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

Here is why SPARK-14437 happens:
BlockManagerId is created using NettyBlockTransferService.hostName which comes from `customHostname`. And `Executor` will set `customHostname` to the hostname which is detected by the driver. However, the driver may not be able to detect the correct address in some complicated network (Netty's Channel.remoteAddress doesn't always return a connectable address). In such case, `BlockManagerId` will be created using a wrong hostname.

To fix this issue, this PR uses `hostname` provided by `SparkEnv.create` to create `NettyBlockTransferService` and set `NettyBlockTransferService.hostname` to this one directly. A bonus of this approach is NettyBlockTransferService won't bound to `0.0.0.0` which is much safer.
## How was this patch tested?

Manually checked the bound address using local-cluster.

### CODE REVIEW

**Comment by zsxwing** [comment]
cc @vanzin

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala [hunk 1]
```diff
@@ -39,7 +39,11 @@ import org.apache.spark.util.Utils
 /**
  * A BlockTransferService that uses Netty to fetch a set of blocks at at time.
  */
-class NettyBlockTransferService(conf: SparkConf, securityManager: SecurityManager, numCores: Int)
+private[spark] class NettyBlockTransferService(
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: @rxin is it a mistake that NettyBlockTransferService was public?

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: Never mind. Just found it's in network package

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
retest this please

**Comment by vanzin** [comment]
LGTM.

**Comment by zsxwing** [comment]
Thanks, Merging to master.

**Pull request closed without merging**